### ISSUE TITLE
Pyright doesn't resolve descriptors when the base class is Any

### ISSUE THREAD

**Issue by zmievsa** [microsoft/pyright#6325]
**Describe the bug**
Whenever the base class of some class resolves to Any, pyright doesn't resolve its attribute descriptors to their real values and instead uses the "not bound to any instance" default.

Sqlalchemy avoids this issue by using the new DeclarativeBase but codebases in the middle of migration to new style might get bitten hard by this bug. 

**Code or Screenshots**
```python
from typing import Any, Generic, Optional, Self, TypeVar, Union, overload
import typing


def declarative_base() -> Any:
    ...


_T_co = TypeVar("_T_co")


class InstrumentedAttribute(Generic[_T_co]):
    pass


class Mapped(Generic[_T_co]):
    if typing.TYPE_CHECKING:

        @overload
        def __get__(self, instance: None, owner: Any) -> InstrumentedAttribute[_T_co]:
            ...

        @overload
        def __get__(self, instance: object, owner: Any) -> _T_co:
            ...

        def __get__(self, instance: Optional[object], owner: Any) -> Union[InstrumentedAttribute[_T_co], _T_co]:
            ...


GoodBase: type[object] = declarative_base()


class Good(GoodBase):
    field: Mapped[str]


reveal_type(Good().field) # str


BadBase = declarative_base()


class Bad(BadBase):
    field: Mapped[str]


reveal_type(Bad().field) # InstrumentedAttribute[str]

```

See how Bad().field is InstrumentedAttribute instead of str? It should be a str too and it is for 1.1.333.

**VS Code extension or command-line**
Both. Command-line version: 1.1.334 (note that this problem does not exist on 1.1.333)

**Comment by erictraut** [comment]
Pyright is working as designed here, so I don't consider this a bug. That said, I understand the challenge this poses, so I'm willing to consider a change in behavior.

The problem is that the first overload of the descriptor's `__get__` method is matched because `Bad` is considered type compatible with `None` (which is the annotated type of the `instance` parameter).

When a class derives from `Any`, it's theoretically type compatible with any other type. That means pyright considers a value of type `Bad` to be type compatible with `None`. However, it should be safe to assume that if the comparison type is an instance of a `@final` class (like `None`), then we can safely assume that a class that derives from `Any` is not type compatible with this comparison type. In other words, pyright can assume that `Bad` is not type compatible with `None`.

I tried making this change, but it creates another problem, in particular, it breaks the `NotImplemented` symbol exported from `builtins.pyi`. It is defined as an instance of `_NotImplementedType` which is a "fake" class that derives from `Any` and is marked `@final`. That means it's affected by the proposed behavior change. The `NotImplemented` symbol is specifically defined in this manner to allow it to be compatible with methods like `__eq__` where the return type is typically annotated as `bool`, but `NotImplemented` is also a valid return value. 

So, I don't see a good way to "fix" this issue without creating other false positives. Let me know if you have any good ideas. For now, I'm going to mark this as "won't fix" because I don't see a way to change the behavior that doesn't involve hacks or non-defensible special cases.

**Comment by zmievsa** [comment]
Big thanks! I keep learning new things through such questions. No, I probably don't know a smarter way to fix it.

**Comment by erictraut** [comment]
I missed the fact that this previously worked on 1.1.333. That was by happenstance only. Pyright previously tracked the `None` type internally using a special data type, but in 1.1.334 I switched the logic to remove this special casing. The goal was to fix a bunch of bugs related to `None`. This is a good example of one such "bug", but in this case the "fix" has this downside.

I'm going to put back the special-case handling specifically for `None` in this case. I don't feel great about this, but it's at least consistent with the older behavior.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.335, which I just published. It will also be included in a future release of pylance.
